File: examples/go/stateful-counter/Counter.tsop.go
package contract

import "tsop"

type Counter struct {
	tsop.StatefulSmartContract
	Count tsop.Bigint // no tag = mutable (stateful)
}

func (c *Counter) Increment() {
	c.Count++
}

func (c *Counter) Decrement() {
	tsop.Assert(c.Count > 0)
	c.Count--
}

File: examples/rust/stateful-counter/Counter.tsop.rs
use tsop::prelude::*;

#[tsop::contract]
pub struct Counter {
    // No #[readonly] = mutable (stateful)
    pub count: Bigint,
}

#[tsop::methods(Counter)]
impl Counter {
    #[public]
    pub fn increment(&mut self) {
        self.count += 1;
    }

    #[public]
    pub fn decrement(&mut self) {
        assert!(self.count > 0);
        self.count -= 1;
    }
}
